### ISSUE TITLE
Implement arithmetic and logical operator support?

### ISSUE THREAD

**Issue by tyarkoni** [nipy/nibabel#939]
This has been suggested before (by me, and probably others), but it would be quite helpful to have nibabel classes (at least in the nifti hierarchy) support arithmetic and logical operations as syntactic sugar for common operations—at least in cases where the semantics can be defined with little or no ambiguity.

E.g., `img1 & img2` would give the logical conjunction of two images; `img1 - img2` the difference; etc. It doesn't seem like users are likely to be confused about what these operators do, and obviously they would only work for operands with identical dimensions/affines.

**Comment by effigies** [comment]
I think we could probably do something like:

```Python
class OperableImage:
    def _op(self, other, op):
        # Check header... need to have axes in the same places
        # Check affine
        # Check shape... maybe some limited broadcasting
        # Check types? Problematic types will be caught by nummpy, but might be cheaper to check before loading data.
        dataobj = op(np.asanyarray(self.dataobj), np.asanyarray(other.dataobj))
        return self.__class__(dataobj, self.affine, self.header)

    __and__ = partial(_op, op=operator.__and__)
    ...
```

This seems like it could easily fit into `DataobjImage`, and delegate out compatibility checks to headers and subclass methods. At least the first class type has to be makeable.

The user definitely gives up control of memory management, e.g., caching, so we'll need to be clear about that tradeoff in the docs.

We also have something like this with `ArraySequence`s in the `streamlines` API:

https://github.com/nipy/nibabel/blob/516434c086b64aae03c69b6549548e5556760002/nibabel/streamlines/array_sequence.py#L58-L89

https://github.com/nipy/nibabel/blob/516434c086b64aae03c69b6549548e5556760002/nibabel/streamlines/array_sequence.py#L459-L506

So perhaps worth thinking about how to turn this into an abstract base class that both objects could derive from.

**Comment by matthew-brett** [comment]
Yes - nice idea.

**Comment by htwangtw** [comment]
I will do it. Any idea where should the code live? Inside `streamlines`?

**Comment by effigies** [comment]
No, `DataobjImage`s shouldn't depend on streamlines. If we use the same tooling, we would want it to be somewhere they could both depend on it. Maybe something like `nibabel/arrayops.py`? We don't have to settle on a name right away.

**Comment by htwangtw** [comment]
Sounds good. We can finalise the names etc later.
